### PR TITLE
perf(occupancy_grid_map_outlier_filter): improve performance

### PR DESCRIPTION
**Pull request by TakanoTaiga**
Signed-off-by: Taiga Takano taiga.takano@tier4.jp

## Description

This PR improves performance with the following major changes:
- Elimination of unnecessary conversions between PCL and ROS Messages.
- Changing the function used for filtering from radiusSearch to nearestKSearch.
- Reducing computational load by revising point cloud processing.

## Measurement
With this PR, the node (measurement is the overall latency of the callback) will be approximately more than twice as fast.

| PERCENTILE | Before [ms] | After [ms] | Improvement Factor |
|-------------|-------------|------------|--------------------|
| 99%         | 13.0        | 6.2        | 2.1 times          |
| 90%         | 10.5        | 4.4        | 2.4 times          |
| 50%         | 0.7         | 0.4        | 1.8 times          |

onOccupancyGridMapAndPointCloud2のcall latency

### CODE REVIEW

**Review by YoshiRi — APPROVED** [review]
LGTM

File: perception/occupancy_grid_map_outlier_filter/src/occupancy_grid_map_outlier_filter_nodelet.cpp [hunk 1]
```diff
@@ -150,43 +150,70 @@ void RadiusSearch2dFilter::filter(
...
     const int points_num =
-      kd_tree_->radiusSearch(i, search_radius_, k_indices, k_distances, min_points_threshold);
+      kd_tree_->nearestKSearch(i, min_points_threshold, k_indices_high, k_distances_high);
 
     if (min_points_threshold <= points_num) {
-      output.points.push_back(low_conf_xyz_cloud.points.at(i));
+      output.points.push_back(outlier.points.at(i));
     } else {
-      outlier.points.push_back(low_conf_xyz_cloud.points.at(i));
+      outlier.points.erase(outlier.points.begin() + i);
```
Comment (veqcc) [inline perception/occupancy_grid_map_outlier_filter/src/occupancy_grid_map_outlier_filter_nodelet.cpp:216]: Does it wotk well?
I think you cannot modify loop target data structures (in this case, `outlier.points` ).

**Comment by veqcc** [comment]
If possible, `.size()` should not be used to many times.
This is because `.size()` 's complexity depends on its data structure (eg. vector is O(1) but linked list is O(N)) and the internal implementation is usually hidden in libraries.

**Comment by YoshiRi** [comment]
@veqcc Is this PR ready for merge?

**Comment by TakanoTaiga** [comment]
@YoshiRi 
The PR has been revised to address certain points that were raised, and is now in a state where it can be merged. As for the changes made to the filter part (changes to the nearestKSearch function), these have been reverted.

**Review by YoshiRi — APPROVED** [review]
Thanks! LGTM

**Comment by TakanoTaiga** [comment]
I measured the latency in the subscriber function onOccupancyGridMapAndPointCloud2 and mainly improved the latency in splitPointCloudFrontBack. The changes made the other day have not improved the speed as initially anticipated.
splitPointCloudFrontBack
onOccupancyGridMapAndPointCloud2

**Comment by veqcc** [comment]
@TakanoTaiga 
Thank you for measuring the performance after improvements!
It looks the first figure indicates that the latency becomes worse, right?

**Comment by veqcc** [comment]
@TakanoTaiga 
You mean the `onOccupancy...` function slightly improved, while the `splitPoint...` function got worse, right?
Your improvements are implemented only in `splitPoint...`, so it sounds strange for me.
Could you give me more detailed explanation?

**Comment by TakanoTaiga** [comment]
まず、`splitPoint...` 関数が `onOccupancy...` 関数内で呼び出されていることに注意する必要があります。そして、`onOccupancy...` 関数はこのノードのメインのコールバックであるため、それを測定します。

メインのコールバックの改善グラフを単に表示するだけでは違いが明確には表示されないかもしれないと考え、改善された `splitPoint...` 関数も表示して、改善をより理解しやすくしました。

**Comment by TakanoTaiga** [comment]
@veqcc 
First, it should be noted that the `splitPoint...` function is called within the `onOccupancy...` function. And since the `onOccupancy...` function is the main callback for this node, it measures it.

I thought that simply displaying the improvement graph of the main callback might not clearly show the difference, so I also displayed the improved `splitPoint...` function to make the improvement more understandable.

**Comment by veqcc** [comment]
@TakanoTaiga 
Thank you! I understand the point.
My question here is "Why `splitPoint...` latency got larger?".

**Comment by TakanoTaiga** [comment]
@veqcc 

I'm sorry. I misunderstood the graph. Tomorrow, I will check if the numbers are incorrect. It seems that I might have confused "before" and "after."

**Comment by TakanoTaiga** [comment]
@veqcc 
As a result of measuring performance to verify once again based on the feedback, performance improvements were observed in both the `splitPoint...` function and the `onOccupancy...` function.

onOccupancyGridMapAndPointCloud2


**onOccupancyGridMapAndPointCloud2**
|  | before[ms] | after[ms] |
|------|--------|-------|
| 99%  | 26.98  | 24.60 |
| 90%  | 21.29  | 18.98 |
| 50%  | 13.88  | 12.77 |


splitPointCloudFrontBack


**splitPointCloudFrontBack**

|   | before[ms] | after[ms] |
|------|--------|-------|
| 99%  | 2.20   | 1.02  |
| 90%  | 1.58   | 0.63  |
| 50%  | 1.07   | 0.38  |

**Review by veqcc — APPROVED** [review]
LGTM :+1:

**Pull request merged**